server/task/src/test/java/org/apache/james/task/MemoryTaskManagerWorkerTest.java

```diff
@@ -0,0 +1,163 @@
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.task;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.junit.Test;
+
+import reactor.core.publisher.Mono;
+
+public class MemoryTaskManagerWorkerTest {
+
+    private final MemoryTaskManagerWorker worker = new MemoryTaskManagerWorker();
+
+    private final Task successfulTask = () -> Task.Result.COMPLETED;
+    private final Task failedTask = () -> Task.Result.PARTIAL;
+    private final Task throwingTask = () -> {
+        throw new RuntimeException("Throwing Task");
+    };
+
+    @Test
+    public void aSuccessfullTaskShouldCompleteSuccessfully() {
+        assertThatTaskSucceed(successfulTask);
+    }
+
+    @Test
+    public void aFailedTaskShouldCompleteWithFailedStatus() {
+        assertThatTaskFailed(failedTask);
+    }
+
+    @Test
+    public void aThrowingTaskShouldCompleteWithFailedStatus() {
+        assertThatTaskFailed(throwingTask);
+    }
+
+    @Test
+    public void theWorkerShouldReportThatATaskIsInProgress() {
+        TaskId id = TaskId.generateTaskId();
+        CountDownLatch latch = new CountDownLatch(1);
+        CountDownLatch taskLaunched = new CountDownLatch(1);
+
+        Task inProgressTask = () -> {
+            taskLaunched.countDown();
+            await(latch);
+            return Task.Result.COMPLETED;
+        };
+
+        TaskWithId taskWithId = new TaskWithId(id, inProgressTask);
+        TaskExecutionDetails executionDetails = TaskExecutionDetails.from(inProgressTask, id);
+        ConcurrentHashMap<TaskId, TaskExecutionDetails> idToDetails = new ConcurrentHashMap<>();
+        idToDetails.put(id, executionDetails);
+
+        worker.executeTask(taskWithId, updateDetails(idToDetails, id)).cache();
+        await(taskLaunched);
+        assertThat(idToDetails.get(id).getStatus()).isEqualTo(TaskManager.Status.IN_PROGRESS);
+        latch.countDown();
+    }
+
+    @Test
+    public void theWorkerShouldNotRunACancelledTask() {
```

Comment:
can you add tests about:
 - submitting a TaskWithId twice? (start it, doesn't wait for finish, start it again)

Reply:
it's more the manager responsability as the worker is stateless.